File: Python/Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param {ListNode} head
    # @return {ListNode}
    def reverseList(self, head):
        if head == None:
            return head
        elif head.next == None:
            return head
        start = head
        temp = head.next
        start.next = None
        while temp.next:
            prev = head
            head = temp
            temp = temp.next
            head.next = prev
        temp.next = head
        return temp

File: cpp/Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == NULL || head->next == NULL)
            return head;
        ListNode *start = head;
        ListNode * temp = head->next;
        ListNode * prev = NULL;

        start->next = NULL;
        while (temp->next) {
            prev = head;
            head = temp;
            temp = temp->next;
            head->next = prev;
        }
        temp->next = head;
        return temp;
    }
};
